Fixes #5556: Feature flags failing to log correctly

kkmurerwa opened the pull request:
## Explanation
Fixes #5556: Feature flags failing to log correctly

### Cause of the bug
The primary cause of the bug was [Google Analytics character limits](https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics.html#logEvent(java.lang.String,android.os.Bundle)) on log event parameter values. Google Analytics (GA), particularly GA 4, imposes the following limits on log events and their parameters;

- **25 Parameters** - Each app event cannot exceed 25 individual parameters. This number is inclusive of default GA parameters making the limit slightly lower than that number.
- **40 Characters** - An event parameter name cannot exceed 40 characters.
- **100 Characters** - Each event parameter can have a value not exceeding 100 characters. If a key or name exceeds the limit, the event is dropped and a `firebase_error` parameter is added to the event and a `firebase_error_value` parameter is added to indicate the name of the parameter with the invalid value. In the case of exceeding the character limit, the error code is usually `4` as can be seen from the list of [GA4 errors](https://firebase.google.com/docs/analytics/errors).

**Note**: These limits only apply for free tier users. Google Analytics 360 users are allowed up to `100 parameters`  per event and `500 characters` for event parameter values.

### Fix implemented
To fix this issue, three changes were made to the `EventBundleCreator`;

- **Feature flag names** - A new converter, the `FeatureFlagNameToIntegerNameConverter` was introduced. This converter will keep a list of flag names supported by the app and their integer representations. Each feature flag name should have a unique integer representation and no two feature flag names should have the same integer representation, even when one of them is no longer supported.
- **Feature flag sync status** - Each feature flag sync status is parsed into an integer representation based on the number it has in the `platform_parameters.proto` that defines the various sync statuses as an enum. The following are the current sync status and their integer representations: `SYNC_STATUS_UNSPECIFIED` as `0`, `NOT_SYNCED_FROM_SERVER` as `1`, and `SYNCED_FROM_SERVER` as `2`.
- **Feature flag enabled state** - Since enabled states are either `true` or `false`, `true` is converted to `1` and `false` to `0`.

By making these changes, it is possible to theoretically log **36 feature flags** in one event. The breakdown for this is as shown;

**Feature flags with 1 digit** - 1 digit each and a comma seperator = _**9 maximum feature flags taking 18 characters**_
**Feature flags with 2 digits** - 2 digits each and a comma seperator = _**27 maximum feature flags taking 81 characters**_
**Total loggable feature flags** = _**9 + 27 = 36 maximum feature flags taking 99/100 characters**_

### Logs comparison
Previously, the feature flags list used to appear like this;

```json
{
    "uuid": null,
    "app_session_id": "277ed911-efef-3144-b0fa-d37cdd87ac86",
    "feature_flag_names": null,
    "feature_flag_enabled_states": "false,false,false,false,false,false,false,false,false,false,true,false,false",
    "feature_flag_sync_statuses": null
}
```

With the new implementation, the feature flags list appears like this;
```json
{
    "uuid": null,
    "app_session_id": "277ed911-efef-3144-b0fa-d37cdd87ac86",
    "feature_flag_sync_statuses": "1,1,1,1,1,1,1,1,1,1,1,1,1",
    "feature_flag_enabled_states": "0,0,0,0,0,0,0,0,0,0,1,0,1",
    "feature_flag_names": "1,2,3,4,5,6,7,8,9,10,11,12,13"
}
```

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

kkmurerwa committed:
Fix feature flags not logging

kkmurerwa committed:
Add type converters for the feature flags log

kkmurerwa committed:
Revert unintended changes on the FeatureFlagsLogger

kkmurerwa committed:
Migrate to passing SyncStatus instead of a string

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagSyncStatusToIntegerConverter.kt:
```diff
@@ -0,0 +1,24 @@
+   * @return An integer representation of the event name.
+   */
+  fun convertToInteger(syncStatus: SyncStatus): Int {
+    return when (syncStatus.toString()) {
+      "SYNC_STATUS_UNSPECIFIED" -> 1
+      "NOT_SYNCED_FROM_SERVER" -> 2
+      "SYNCED_FROM_SERVER" -> 3
+      else -> -1
+    }
```
This should just `when` on the enum directly since then no conversion happens (and it can be exhaustive so `else` can be omitted).

BenHenning replied:
Also, could we maybe just use the SyncStatus's number or index field that proto generates? That is a stable integer representation of the enum.

kkmurerwa replied:
NTS: Use `syncStatus.number` to get the index

kkmurerwa replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerConverter.kt:
```diff
@@ -0,0 +1,46 @@
+      ENABLE_NPS_SURVEY -> 11
+      ENABLE_ONBOARDING_FLOW_V2 -> 12
+      ENABLE_MULTIPLE_CLASSROOMS -> 13
+      else -> -1
```
Let's use zero to avoid negatives for the error case.

kkmurerwa replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:
```diff
@@ -688,12 +688,19 @@ class EventBundleCreator @Inject constructor(
+          featureFlagSyncStatusConverter.convertToInteger(it.flagSyncStatus)
+        }
+
         val featureFlagEnabledStates = featureFlagsList.map { it.flagEnabledState }
```
As discussed, this should be replaced with 1s and 0s for trues and falses.

kkmurerwa replied:
NTS: Use `syncStatus.compareTo(false)`

kkmurerwa replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerConverter.kt:
```diff
@@ -0,0 +1,46 @@
+/**
+ *
+ */
+class FeatureFlagNameToIntegerConverter @Inject constructor() {
```
Doesn't need to be injectable and can just be an object to avoid instantiation.

BenHenning replied:
Also, let's use 'NumericId' instead of 'Integer' here and below.

kkmurerwa replied:
Done

BenHenning replied:
The 'NumericId' bit is still unresolved--please fix.

kkmurerwa replied:
This should be resolved now.

kkmurerwa committed:
Improve how feature flag sync statuses and enabled states are converted to integers

kkmurerwa committed:
Add tests for the FeatureFlagToIntegerNameConverter class

kkmurerwa committed:
Add FlagNameToIntegerNameConverterTest.kt to list of exempted parameterized tests

kkmurerwa commented:
@adhiamboperes this is ready for another pass. PTAL

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverter.kt:
```diff
@@ -0,0 +1,46 @@
+import org.oppia.android.util.platformparameter.SPOTLIGHT_UI
+import javax.inject.Inject
+
+/**
+ *
+ */
```
Incomplete KDoc.

kkmurerwa replied:
I have added text here. I have also updated the Wiki so that contributors can know to add new feature flags to the flag name to integer converter.

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverter.kt:
```diff
@@ -0,0 +1,46 @@
+  /**
+   * Converts a feature flag name to an integer.
+   *
+   * @param flagName The flag name to convert.
+   * @return An integer representation of the event name.
```
```suggestion
    * @param flagName the string constant flag name to convert.
   * @return an integer representation of the flag name.
```

adhiamboperes replied:
Per last weeks discussion, we don't want to re-use the integers. We should make a note of that in the KDoc.

kkmurerwa replied:
Done

adhiamboperes commented on utility/src/test/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverterTest.kt:
```diff
@@ -0,0 +1,154 @@
+    "expectedValue=13"
+  )
+  fun testConvertToIntegerName_returnsCorrectIntegerForEach() {
+    val nameInteger = converter.convertToInteger(flagName)
```
Maybe?
```suggestion
    val integerName = converter.convertToInteger(flagName)
```

kkmurerwa replied:
I have implemented this suggestion.

adhiamboperes reviewed (changes requested):
Thanks @kkmurerwa! Implementation looks good, but the PR description could benefit from a thorough write up since we havent documented the character length issues elsewhere (AFAIK) and why we chose this solution. The description could also benefit from sample logs as copied over from println debug and Firebase logs.

kkmurerwa committed:
Refine KDocs for the FeatureFlagNameToIntegerNameConverter class

kkmurerwa committed:
Add Wiki documentation for the FeatureFlagNameToIntegerNameConverter

kkmurerwa committed:
Merge branch 'develop' into fix_feature_flags_not_logging

kkmurerwa commented:
@adhiamboperes I have updated the PR description and addressed the other comments. PTAL

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverter.kt:
```diff
@@ -16,14 +16,19 @@ import org.oppia.android.util.platformparameter.SPOTLIGHT_UI
 
 /**
- *
+ * Converter for feature flag names into integers for more compact logging to address Google
```
```suggestion
 * Converter for feature flag names to integers for more compact logging to address Google
```

kkmurerwa replied:
Done

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverter.kt:
```diff
@@ -16,14 +16,19 @@ import org.oppia.android.util.platformparameter.SPOTLIGHT_UI
 /**
- *
+ * Converter for feature flag names into integers for more compact logging to address Google
+ * Analytics character limits. GA4 limits the characters permitted in a log event parameter value to
+ * a maximum of 100 characters as of March 2025.
```
I believe that when we include information like this in KDocs, the standard pattern is to add the url(e.g. https://github.com/oppia/oppia-android/blob/9404ea41d91d0a2a62ca8d42de29d60d234fd02c/app/src/main/java/org/oppia/android/app/application/AbstractOppiaApplication.kt#L45-L47)

kkmurerwa replied:
I wanted to do this but the link is too long and exceeds the 100 character limit.

BenHenning replied:
You should be able to include it without triggering a lint failure. It's better to include the link for context even if it goes over 100 characters.

kkmurerwa replied:
I have included it.

BenHenning replied:
Marking this as resolved since I've verified it (feel free to reopen this if you'd like @adhiamboperes).

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverter.kt:
```diff
@@ -16,14 +16,19 @@ import org.oppia.android.util.platformparameter.SPOTLIGHT_UI
    *
-   * @param flagName The flag name to convert.
-   * @return An integer representation of the event name.
+   * These numbers are supposed to be increment indefinitely and the integer representation of
```
```suggestion
   * These numbers are supposed to be incremented indefinitely and the integer representation of
```

kkmurerwa replied:
Done

adhiamboperes reviewed (changes requested):
Thanks @kkmurerwa! I left some further suggestions for KDocs, but minor. Also, it would be great to have a sample log of the current results e.g:

```json
[{
  "uuid": null,
  "app_session_id": "277ed911-efef-3144-b0fa-d37cdd87ac86",
  "feature_flag_names": null,
  "feature_flag_enabled_states": "false,false,false,false,false,false,false,false,false,false,true,false,false",
  "feature_flag_sync_statuses": null
}, {
  "uuid": null,
  "app_session_id": "f2ed5da7-bd3d-3ec3-9044-13ddf766d8fe",
  "feature_flag_names": null,
  "feature_flag_enabled_states": "false,false,false,false,false,false,false,false,false,false,true,false,true",
  "feature_flag_sync_statuses": null
}, {
  "uuid": null,
  "app_session_id": "a41e3a56-4c4a-3a3b-b820-a6337dd5797c",
  "feature_flag_names": null,
  "feature_flag_enabled_states": "false,false,false,false,false,false,false,false,false,false,true,false,true",
  "feature_flag_sync_statuses": null
}]
``` 

Since as reviewers we don't have other means of verifying the wxpected output.

kkmurerwa committed:
Merge branch 'develop' into fix_feature_flags_not_logging

kkmurerwa committed:
Minor refactors to the KDocs on FeatureFlagNameToIntegerNameConverter

kkmurerwa commented:
@adhiamboperes I have updated the PR description to include the logs. I have also added the changes suggested with the exception of the link which is too long and exceeds the character limit. PTAL

kkmurerwa committed:
Merge branch 'develop' into fix_feature_flags_not_logging

adhiamboperes reviewed (approved):
Thanks @kkmurerwa! I have no further comments. Assigning @BenHenning for verification and resolving open comments.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:
```diff
@@ -688,12 +688,15 @@ class EventBundleCreator @Inject constructor(
+        }
+
+        val featureFlagSyncStatuses = featureFlagsList.map { it.flagSyncStatus.number }
+        val featureFlagEnabledStates = featureFlagsList.map { it.flagEnabledState.compareTo(false) }
```
`compareTo` is an interesting approach, though it's possibly going to be hard for people to understand why it works. It also supports a theoretical tri-state: less than (negative), equal (0), greater than (positive). I suggest just doing a basic integer conversion, instead:

```suggestion
        val featureFlagEnabledStates = featureFlagsList.map { if (it) 1 else 0 }
```

kkmurerwa replied:
I have changed the approach to this.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:
```diff
@@ -688,12 +688,15 @@ class EventBundleCreator @Inject constructor(
-        val featureFlagNames = featureFlagsList.map { it.flagName }
-        val featureFlagSyncStatuses = featureFlagsList.map { it.flagSyncStatus }
-        val featureFlagEnabledStates = featureFlagsList.map { it.flagEnabledState }
+        val featureFlagNameConverter = FeatureFlagNameToIntegerNameConverter()
```
Either inject the converter, or convert it to an object (not a class). In neither case should we be instantiating it here.

kkmurerwa replied:
I converted it to an object

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverter.kt:
```diff
@@ -0,0 +1,51 @@
+   * These numbers are supposed to be incremented indefinitely and the integer representation of
+   * each feature flag name should not be reused even after the feature flag is no longer in use.
+   *
+   * @param flagName the string constant flag name to convert.
+   * @return an integer representation of the flag name.
```
```suggestion
   * @param flagName the string constant flag name to convert
   * @return an integer representation of the flag name
```

Nit: we don't end at-clauses with punctuation per style guide.

kkmurerwa replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverter.kt:
```diff
@@ -0,0 +1,51 @@
+   */
+  fun convertToInteger(flagName: String): Int {
+    return when (flagName) {
+      DOWNLOADS_SUPPORT -> 1
+      EXTRA_TOPIC_TABS_UI -> 2
+      LEARNER_STUDY_ANALYTICS -> 3
+      FAST_LANGUAGE_SWITCHING_IN_LESSON -> 4
+      LOGGING_LEARNER_STUDY_IDS -> 5
+      EDIT_ACCOUNTS_OPTIONS_UI -> 6
+      ENABLE_PERFORMANCE_METRICS_COLLECTION -> 7
+      SPOTLIGHT_UI -> 8
+      INTERACTION_CONFIG_CHANGE_STATE_RETENTION -> 9
+      APP_AND_OS_DEPRECATION -> 10
+      ENABLE_NPS_SURVEY -> 11
+      ENABLE_ONBOARDING_FLOW_V2 -> 12
+      ENABLE_MULTIPLE_CLASSROOMS -> 13
```
Can you please update these to number each feature based on when its flag was first created (i.e. they should be date sorted)? This will make it seem as though these were always used (as if we had used the system since the first feature flag was added), and ensures there's always a correspondence between the integer value and the age of the feature (i.e. that a larger number always indicates a newer feature).

I'll be using these for the new feature flag indexes in #5725 (which will allow us to remove this class actually since then we can just use the proto enum indexes).

kkmurerwa replied:
I tracked the order in which they were created and documented it in this [Google Sheet](https://docs.google.com/spreadsheets/d/1MbW-O3toEUQu1LUXozDkYZCfBi8kDn2jN86nUwTWWLA/edit?usp=sharing). I have updated this block of code based on that list. Let me know if I should include feature flags that have already been removed.

BenHenning replied:
We should account for the removed ones, and I did an analysis as well (before seeing this link, though I also need to know platform parameters for my PR). Left a new comment accordingly.

BenHenning commented on wiki/Platform-Parameters-&-Feature-Flags.md:
````diff
@@ -217,6 +217,28 @@ Besides the feature-flag logger, the `FeatureFlagLoggerTest` located at `domain/
+}
+```
+
+**DO NOT** reuse any of the digits that have already been used. The numbers are supposed to increment to infinity.
````
```suggestion
**DO NOT** reuse any of the digits that have already been used.
```

Not sure the second bit adds anything. We could alternatively say "The upper bound is unexpected to be hit." or something, but I'm not sure that's really needed either. Am I maybe missing something?

kkmurerwa replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt:
```diff
@@ -688,12 +688,15 @@ class EventBundleCreator @Inject constructor(
-        val featureFlagEnabledStates = featureFlagsList.map { it.flagEnabledState }
+        val featureFlagNameConverter = FeatureFlagNameToIntegerNameConverter()
+
+        val featureFlagNames = featureFlagsList.map {
+          featureFlagNameConverter.convertToInteger(it.flagName)
+        }
+
+        val featureFlagSyncStatuses = featureFlagsList.map { it.flagSyncStatus.number }
+        val featureFlagEnabledStates = featureFlagsList.map { it.flagEnabledState.compareTo(false) }
```
Why are these changes not leading to failures in `EventBundleCreator`? It seems that perhaps we are missing coverage--can you please make sure these changes are properly tested?

kkmurerwa replied:
I have added some tests to the EventBundleCreatorTest. I chose to go with a list of feature flag item contexts instead of parameterized tests because it allows us to better test the logs in the exact format they will be in.

BenHenning reviewed (changes requested):
Thanks @kkmurerwa! I was thinking that we should also add a top-level converter test to verify the count of flags, but I'm not too bothered since #5725 will actually largely simplify this PR's approach since proto enums will provide the exact same monotonic property without needing to actually test it in the same way.

Separately, can you add an `EventBundleCreator` test that fails if any of the feature flag parameter values exceed 100 characters? And, please verify that this fails without your fixes in place. Not only does it prove the fix, it also provides a gate for when we inevitably hit this in the future when we have more than ~30 active features at once (or ~25 if we hit 3 digit numbers).

kkmurerwa committed:
Make minor changes from peer review

kkmurerwa committed:
Merge branch 'develop' into fix_feature_flags_not_logging

kkmurerwa committed:
Change FeatureFlagNameToIntegerNameConverter from a class to an object

kkmurerwa committed:
Merge remote-tracking branch 'origin/develop' into fix_feature_flags_not_logging

kkmurerwa committed:
Reorganize the feature flag names list into chronological order

kkmurerwa committed:
Add tests to ensure the EventBundleCreator correctly parses feature flag events

kkmurerwa committed:
Refactor FeatureFlagNameToIntegerNameConverter tests to align with the new numbering scheme

kkmurerwa commented:
@BenHenning I have addressed the comments. PTAL

BenHenning commented on utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt:
```diff
@@ -848,6 +852,27 @@ class EventBundleCreatorTest {
   }
 
+  @Test
+  fun testFillEventBundle_featureFlagsContext() {
```
```suggestion
  fun testFillEventBundle_fillsFeatureFlagsContext() {
```

The expected outcome should be documented as an action.

kkmurerwa replied:
Done

BenHenning commented on utility/src/test/java/org/oppia/android/util/logging/EventBundleCreatorTest.kt:
```diff
@@ -848,6 +852,27 @@ class EventBundleCreatorTest {
+    assertThat(bundle).integer("android_sdk").isEqualTo(TEST_ANDROID_SDK_VERSION)
+    assertThat(bundle).string("app_version_name").isEqualTo(TEST_APP_VERSION_NAME)
+    assertThat(bundle).integer("app_version_code").isEqualTo(TEST_APP_VERSION_CODE)
+    assertThat(bundle).string("feature_flag_names").isEqualTo("0,0,0,7")
```
This feels a bit fake--we should never realistically logging the same flag multiple times, right? Can you please make the test match a more realistic case (i.e. different flags)?

kkmurerwa replied:
I have updated this to include actual feature flag names as they appear. Let me know if this is better.

BenHenning commented on utility/src/main/java/org/oppia/android/util/logging/FeatureFlagNameToIntegerNameConverter.kt:
```diff
@@ -13,35 +13,36 @@ import org.oppia.android.util.platformparameter.INTERACTION_CONFIG_CHANGE_STATE_
+      LOGGING_LEARNER_STUDY_IDS -> 10
       ENABLE_NPS_SURVEY -> 11
       ENABLE_ONBOARDING_FLOW_V2 -> 12
       ENABLE_MULTIPLE_CLASSROOMS -> 13
```
This order still doesn't seem right to me, so I did an analysis just to double check and to also account for platform parameters (for my PR).

- 19 Jul 2024: remove ENABLE_LANGUAGE_SELECTION_UI_DEFAULT_VALUE
- 28 May 2024: add multiple classrooms (feature)
- 06 Feb 2024: add onboarding v2 (feature)
- 25 Jan 2024: add nps (feature)
- 16 Nov 2023: remove continue button animation
- 22 Jun 2023: add npcsurveygraceperiod (parameter), npssurveyminimaggregatelearningtimeinatopic (parameter), 
- 05 Jun 2023: add fast language switching (feature), logging learner study ids (feature)
- 26 May 2023: add os deprecation (feature), app update version code (parameter), force update version code (parameter), lowest api level (parameter)
- 19 Nov 2022: add continue button animation (platform parameter, but probably actually a feature)
- 10 Nov 2022: rename AutomaticUpdateTopicSetting to EnableDownloadsSupport
- 18 Oct 2022: add EnableInteractionConfigChangeStateRetention (feature)
- 03 Oct 2022: add AutomaticUpdateTopicSetting (feature)
- 27 Sep 2022: add extra topics UI (feature)
- 24 Sep 2022: add spotlights (feature)
- 18 Jul 2022: add enable edit account options (feature)
- 08 Jun 2022: enable performance metrics collection (feature), perfmetcollectionuploadtime (parameter), perfmetricscollectionhighfreq (parameter), perfmetricscollectionlowfreq (parameter)
- 26 Mar 2022: add cachelatexrendering (parameter)
- 24 Jan 2022: add learner study analytics (feature)
- 06 Oct 2021: add enable language selection UI (feature)
- 20 Aug 2021: add sync up worker time period (parameter)
- 08 Jul 2021: add splash screen welcome message (parameter)

Complete list of parameters (from oldest to newest):
- `SplashScreenWelcomeMsg`
- `SyncUpWorkerTimePeriodHours`
- `CacheLatexRendering`
- `PerformanceMetricsCollectionUploadTimeIntervalInMinutes`
- `PerformanceMetricsCollectionHighFrequencyTimeIntervalInMinutes`
- `PerformanceMetricsCollectionLowFrequencyTimeIntervalInMinutes`
- `OptionalAppUpdateVersionCode`
- `ForcedAppUpdateVersionCode`
- `LowestSupportedApiLevel`
- `NpsSurveyGracePeriodInDays`
- `NpsSurveyMinimumAggregateLearningTimeInATopicInMinutes`

Complete list of features (from oldest to newest, deleted ones indicated):
- REMOVED: `EnableLanguageSelectionUi`: expected numeric value of 1
- `EnableLearnerStudyAnalytics`: expected numeric value of 2
- `EnablePerformanceMetricsCollection`: expected numeric value of 3
- `EnableEditAccountsOptionsUi`: expected numeric value of 4
- `EnableSpotlightUi`: expected numeric value of 5
- `EnableExtraTopicTabsUi`: expected numeric value of 6
- `EnableDownloadsSupport` (previously: `AutomaticUpdateTopicSetting`): expected numeric value of 7
- `EnableInteractionConfigChangeStateRetention`: expected numeric value of 8
- REMOVED: `EnableContinueButtonAnimation`: expected numeric value of 9
- `EnableAppAndOsDeprecation`: expected numeric value of 10
- `EnableFastLanguageSwitchingInLesson`: expected numeric value of 11
- `EnableLoggingLearnerStudyIds`: expected numeric value of 12
- `EnableNpsSurvey`: expected numeric value of 13
- `EnableOnboardingFlowV2`: expected numeric value of 14
- `EnableMultipleClassrooms`: expected numeric value of 15

Looks like it was close; downloads and interaction config change were swapped (probably due to downloads having a different name originally), and 2 removed flags weren't accounted for (which they need to be because we'll be reserving those old values so that they specifically don't get reused or, rather, ever used).

kkmurerwa replied:
I have swapped the `EnableInteractionConfigChangeStateRetention` and `EnableDownloadsSupport` flag names to  the new numeric IDs and renumbered everything else to reflect the suggestion.

BenHenning reviewed (changes requested):
Thanks @kkmurerwa. Left a few new comments, though almost everything is resolved.

However, I think you didn't fully address this part of my previous review's top-level comment:

> Separately, can you add an `EventBundleCreator` test that fails if any of the feature flag parameter values exceed 100 characters? And, please verify that this fails without your fixes in place. Not only does it prove the fix, it also provides a gate for when we inevitably hit this in the future when we have more than ~30 active features at once (or ~25 if we hit 3 digit numbers).

Please make sure to.

kkmurerwa committed:
Merge branch 'develop' into fix_feature_flags_not_logging

kkmurerwa committed:
Merge branch 'develop' into fix_feature_flags_not_logging

kkmurerwa committed:
Rename FeatureFlagNameToIntegerNameConverter to FeatureFlagNameToNumericIdConverter

kkmurerwa committed:
Renumber the feature flag name to id mapping to account for removed and renamed feature flags

kkmurerwa committed:
Update the EventBundleCreatorTest to make feature flag tests more comprehensive

kkmurerwa committed:
Fix oppia-parameterized runner checks

adhiamboperes committed:
Merge branch 'develop' into fix_feature_flags_not_logging

adhiamboperes commented:
Updated branch to retrigger checks - did not find any stacktraces from the stuck run.

adhiamboperes commented:
@kkmurerwa, aside from the failing code coverage failures, You have failing tests: https://github.com/oppia/oppia-android/actions/runs/14720814671/job/41314244218?pr=5715#step:12:328

kkmurerwa committed:
Fix failing test on the FeatureFlagNameToNumericIdConverterTest class

kkmurerwa commented:
@BenHenning this is now ready for another pass. PTAL

kkmurerwa committed:
Merge branch 'develop' into fix_feature_flags_not_logging

kkmurerwa committed:
Merge branch 'develop' into fix_feature_flags_not_logging

BenHenning reviewed (approved):
Thanks @kkmurerwa. Really sorry for the long review delay here, and I appreciate your patience.

This LGTM!
